src/BizHawk.Client.EmuHawk/CustomControls/ClipboardEventTextBox.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Windows.Forms;
+
+namespace BizHawk.Client.EmuHawk.CustomControls
+{
+	public class ClipboardEventTextBox : TextBox
+	{
+		protected override void WndProc(ref Message m)
+		{
+			const int WM_PASTE = 0x302;
+
+			if (m.Msg == WM_PASTE)
```

Comment:
This still needs to be tested on Linux. I hope Mono emulates this window message?

Reply:
Doesn't work :(

Printing `m.Msg` while doing focus, Ctrl+V, defocus gives: 0x00F, 0x085, 0x007, 0x100, 0x100, 0x102, 0x101, 0x101, 0x008

Reply:
Just focus/paint/key messages it seems.

Could fall back to trapping Ctrl+V on Linux, I suppose.

The benefit of `WM_PASTE` is that it also works when pasting through the native right click menu on Windows.

Reply:
I still don't understand why this Win32 stuff was used when there's a `ValueChange` event. You can even detect copy-paste by checking if the Levenshtein distance between the current and pending values is > 1, though I'm not sure that's necessary.

Reply:
As it is, doing this in a change event is too late because the pasted text will have been truncated by the `MaxLength` already, so if a user pastes `0x1234` you just get `0x12` and can't recover the full text. Trying to indirectly infer copy-paste seems much more hacky and fragile to me.

Of course you could go with Morilli's suggestion and overhaul the controls completely. Get rid of all the input filtering and max length, let the user enter whatever they want, and then write new validation/sanitation logic in the `Validating` event or something.